File: shorturl/src/main/java/com/yilong/shorturl/util/Number62.java
package com.yilong.shorturl.util;

public class Number62 {
    public static String ALPHABET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
    private static final int BASE = 62;

    public static String encode(long num) {
        if (num < 0) {
            throw new IllegalArgumentException("num must be greater than 0.");
        }

        StringBuilder sb = new StringBuilder();
        while (num > 0) {
            sb.append(ALPHABET.charAt((int) (num % BASE)));
            num /= BASE;
        }

        return (sb.length() == 0) ? ALPHABET.substring(0, 1) : sb.reverse().toString();
    }

    public static long decode(String str) {
        if (str == null || str.length() == 0 ){
            throw new IllegalArgumentException("str must not be empty.");
        }

        long result = 0;
        int len = str.length();
        int alphabetIndex;
        for (int i = 0; i < len; i++) {
            alphabetIndex = ALPHABET.indexOf(str.charAt(len - 1 - i));
            if (alphabetIndex < 0) {
                throw new IllegalArgumentException("str char must be alphabet.");
            }
            result += alphabetIndex * Math.pow(BASE, i);
        }

        return result;
    }

    private Number62(){}
}

File: fullstack/src/test/utils/number62.ts
import { equal, ok } from 'assert';
import number62 from '../../utils/number62';

describe('number62', function () {

  describe('#encode()', function () {
    it('number62.encode(36) should equal "a"', function () {
      equal(number62.encode(36), 'a');
    });
    it('number62.encode(62) should equal "10"', function () {
      equal(number62.encode(62), '10');
    });
    it('number62.encode(Math.pow(62,2)) should equal "100"', function () {
      equal(number62.encode(Math.pow(62, 2)), '100');
    });
    it('number62.encode(Math.pow(62,2),7) should equal "0000100"', function () {
      equal(number62.encode(Math.pow(62, 2), 7), '0000100');
    });
  });

  describe('#decode()', function () {
    it('number62.decode(a) should equal 36', function () {
      equal(number62.decode('a'), 36);
    });
    it('number62.decode(10) should equal 62', function () {
      equal(number62.decode('10'), 62);
    });
    it('number62.decode(100) should equal Math.pow(62,2)', function () {
      equal(number62.decode('100'), Math.pow(62, 2));
    });
    it('number62.decode(0000100) should equal Math.pow(62,2)', function () {
      equal(number62.decode('0000100'), Math.pow(62, 2));
    });
  });

});